presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.DeleteNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.TableFinishNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+/**
+ * If the predicate for a delete is optimized to false, the target table scan
+ * of the delete will be replaced with an empty values node. This type of
+ * plan cannot be executed and is meaningless anyway, so we replace the
+ * entire thing with a values node.
+ */
+public class RemoveEmptyDelete
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+    {
+        if (!(node instanceof TableFinishNode)) {
```

Comment:
If you're into some 'nasty, functional, no more ifs' code, have a look at:
https://github.com/Teradata/presto/commit/250c0155323a247af6edddbac9115910d8765526
https://github.com/Teradata/presto/commit/bec2b021507a593664f980f208284f3599459286
https://github.com/Teradata/presto/commit/9062e93aaed39e2f2c8971e813435c4247a093ff

also https://github.com/Teradata/presto/commits/feature/rule_migrations

cc @kokosing